File: packages/react-native-nitro-markdown/ios/HybridMarkdownSession.swift
import Foundation
import NitroModules

class HybridMarkdownSession: HybridMarkdownSessionSpec {
    private var buffer = ""
    private var listeners: [UUID: () -> Void] = [:]
    private let lock = NSLock()
    
    private(set) var version: Int = 0
    
    var highlightPosition: Double = 0
    

    
    var memorySize: Int {
        return buffer.utf8.count + MemoryLayout<HybridMarkdownSession>.size
    }
    
    func append(chunk: String) throws {
        lock.lock()
        buffer += chunk
        version += 1
        lock.unlock()
        notifyListeners()
    }
    
    func clear() throws {
        lock.lock()
        buffer = ""
        highlightPosition = 0
        version += 1
        lock.unlock()
        notifyListeners()
    }
    
    func getAllText() throws -> String {
        lock.lock()
        defer { lock.unlock() }
        return buffer
    }
    
    func addListener(listener: @escaping () -> Void) throws -> () -> Void {
        let id = UUID()
        lock.lock()
        listeners[id] = listener
        lock.unlock()
        
        return { [weak self] in
            self?.lock.lock()
            self?.listeners.removeValue(forKey: id)
            self?.lock.unlock()
        }
    }
    
    private func notifyListeners() {
        lock.lock()
        let currentListeners = Array(listeners.values)
        lock.unlock()
        
        for listener in currentListeners {
            listener()
        }
    }
}

File: packages/react-native-nitro-markdown/android/src/main/java/com/margelo/nitro/com/nitromarkdown/HybridMarkdownSession.kt
package com.margelo.nitro.com.nitromarkdown

class HybridMarkdownSession : HybridMarkdownSessionSpec() {
    private var buffer = StringBuilder()
    private val listeners = mutableMapOf<Long, () -> Unit>()
    private var nextListenerId = 0L
    private val lock = Any()

    override var highlightPosition: Double = 0.0
        set(value) {
            synchronized(lock) { field = value }
            // No notify for highlighting to avoid flood
        }



    override val memorySize: Long
        get() = buffer.length.toLong()

    override fun append(chunk: String) {
        synchronized(lock) {
            buffer.append(chunk)
        }
        notifyListeners()
    }

    override fun clear() {
        synchronized(lock) {
            buffer.clear()
            highlightPosition = 0.0
        }
        notifyListeners()
    }

    override fun getAllText(): String {
        synchronized(lock) {
            return buffer.toString()
        }
    }

    override fun addListener(listener: () -> Unit): () -> Unit {
        val id: Long
        synchronized(lock) {
            id = nextListenerId++
            listeners[id] = listener
        }
        return {
            synchronized(lock) {
                listeners.remove(id)
            }
        }
    }

    private fun notifyListeners() {
        val currentListeners: Collection<() -> Unit>
        synchronized(lock) {
            currentListeners = listeners.values.toList()
        }
        currentListeners.forEach { it() }
    }
}
